---
title: "scPipe: flexible data preprocessing pipeline for 3' end scRNA-seq data"
author: "Luyi Tian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scPipe: flexible data preprocessing pipeline for 3' end scRNA-seq data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



# Get started

`scPipe` is a package designed to process scRNA-seq data generated by different protocols. It is designed for protocols with UMI, but can also adapt to non-UMI protocols. scPipe consist of two major components. The first is data preprocessing with raw fastq as input and gene counting matrix as output. The second component starts from gene counting matrix, includes quality control and a shiny app for clustering and visualization.

# case study: process CEL-seq2 data

## a simple CEL-seq2 dataset.

we start from creating a folder to store all results.

```{r,eval=TRUE}
library(scPipe)
library(SingleCellExperiment)
data_dir = tempdir()
```

CEL-seq2 is a modified version of CEL-Seq method with higher sensitivity and lower costs. It uses a polyT primer with index sequences to capture cell's mRNA. The index sequence consist of the a cellular specific index, which is designed and the same for each cell, and a molecular index, also called unique molecular identifier (UMI), which is randomly generated and different for each primer. The ouput fastq of Cel-seq2 is paired-ended where the first read is transcript and second read is cellindex+UMI with some polyA sequence.

To process the data, we need the genome fasta file, gff3 exon annotation and a cell barcode annotation. The barcode annotation should be a `.csv` file with at least have two column where first column is the cell id and second column is the barcode sequence. We use ERCC in this demo. All file can be found in `extdata` in this package:

```{r}
# file path:
ERCCfa_fn = system.file("extdata", "ERCC92.fa", package = "scPipe")
ERCCanno_fn = system.file("extdata", "ERCC92_anno.gff3", package = "scPipe")
barcode_annotation_fn = system.file("extdata", "barcode_anno.csv", package = "scPipe")
```

We start from the paired-end fastq files. Read1 contains transcript and Read2 contains UMI followed by cell barcode.  The read structure for CEL-seq2 is a paired-ended reads with first longer read mapped to transcript and second shorter read consist of 6bp UMI followed by 8bp cell barcode. **NOTE**: by convention, in scPipe the read1 always refer to the reads with transcript sequence, usually the longer reads.
```{r,eval=TRUE}
fq_R1 = system.file("extdata", "simu_R1.fastq.gz", package = "scPipe")
fq_R2 = system.file("extdata", "simu_R2.fastq.gz", package = "scPipe")
```

## data proprocessing

The pipeline start from fastq file reformatting. We move the barcode and UMI to the read name and keep the transcript, so the output read name looks like `@[barcode_sequence]*[UMI_sequence]#[readname]` ...

```{r,eval=TRUE}
sc_trim_barcode(file.path(data_dir, "combined.fastq"),
                fq_R1,
                fq_R2)
```

Then we align the reads to genome. The example uses `Rsubread` but it can be any aligner that support RNA-seq alignment and gives standard bam output. The `Rsubread` package does not support Windows platform so the following step will not run on Windows.

```{r,eval=TRUE}
if(!(.Platform$OS.type == "windows")){
  Rsubread::buildindex(basename=file.path(data_dir, "ERCC_index"), reference=ERCCfa_fn)

  Rsubread::align(index=file.path(data_dir, "ERCC_index"),
      readfile1=file.path(data_dir, "combined.fastq"),
      output_file=file.path(data_dir, "out.aln.bam"), phredOffset=64)
}

```

After the read alignment we will map reads to exons using the exon annotations. Currently `scPipe` only supports annotation in `gff3` and `bed` format.

```{r,eval=TRUE}
if(!(.Platform$OS.type == "windows")){
  sc_exon_mapping(file.path(data_dir, "out.aln.bam"),
                file.path(data_dir, "out.map.bam"),
                ERCCanno_fn)
}

```

Next we use `sc_demultiplex()` to split the reads to each cell and put them into spearate `.csv` file in `/count` subfolder where each file contains three columns and each row is a read with first column being the gene that the read mapped to, second column is the UMI sequence of this read, the third column is the distance to transcript end position (TES). This file will be used for UMI deduplication and to generate gene counting matrix by callling `sc_gene_counting()`.

```{r,eval=TRUE}
if(!(.Platform$OS.type == "windows")){
  sc_demultiplex(file.path(data_dir, "out.map.bam"), data_dir, barcode_annotation_fn,has_UMI=FALSE)

  sc_gene_counting(data_dir, barcode_annotation_fn)
}

```


Now we complete the preprocessing step. you will find the gene counting matrix at `data_dir/gene_count.csv` and quality control statistics in `data_dir/stat` folder, which can be used in the following step for quality control.


# process data generated by other protocols

the most difference that come from different protocols is the read structure, which is defined by `read_structure` argument in scPipe. The difference between CEL-seq and Drop-seq is the cell barcode is unknown for Drop-seq, so an extra step `sc_detect_bc` is required before `sc_demultiplex`, to identify and generate the cell barcode annotation. To use `run_scPipe` on Drop-seq data, just ignore the `barcode_anno` argument. scPipe is not optimized for the full-lengh protocols like SMART-seq, but it can process the data generated by such protocol by set the `has_UMI` to `FALSE` in `sc_demultiplex` or `run_scPipe`.

## quality control

The easiest way to create a SingleCellExperiment object is using `create_sce_by_dir`, it will read the gene counting matrix as well as the quality control informations in stat folder.

```{r}
sce = create_sce_by_dir(data_dir)
dim(sce)
```

The dataset we created using ERCC simulated reads only contains 10 cells with perfect quality. So we will use the example dataset in scPipe, which is a part from Christine Biben's blood cells with different cell type that contains 383 cells and 1000 genes, to show the quality control functions in scPipe. This shows the way to create the SingleCellExperiment object without using `create_sce_by_dir`:

```{r}
data("sc_sample_data")
data("sc_sample_qc")
sce = SingleCellExperiment(assays = list(counts =as.matrix(sc_sample_data))) # generate new sce with gene count matrix
QC_metrics(sce) = sc_sample_qc
demultiplex_info(sce) = cell_barcode_matching
UMI_dup_info(sce) = UMI_duplication

```


There are several plots we can create to access the overall quality of the experiment. The first is cell barcode demultiplex statistics. The pie chart shows the percentage of reads that uniquely match a cell barcodes.It also gives, for the unmathced part, their alignmen rate and the percentage of reads mapped to exon. The large proportion of unmatch read mapped to exon indicate a failure of cell barcode demultiplexing.

```{r}
plot_demultiplex(sce)
```

Another plot, which usually used to evaluate the read deapth, is the duplication number of UMIs. The UMI is used to mark each individual molecules and after PCR amplification, the same molecule with have multiple copys, which can be identified and removed by the same UMI sequence. Therefore, the copy number of UMIs is an indication of the PCR amplification rate. 

```{r}
plot_UMI_dup(sce)
```

Then we calculate the QC metrics and use the `detect_outlier` function to find the poor quality cells. The `detect_outlier` has argument `comp` to define the maximum component of the gaussian mixture model. Usually the default value 1 is sufficient, but in case the data is heterogeneous in terms of quality control metrics, the value can be set to 2 or 3 to get a better results. It can only remove low quality cells by `type="low"` or only cells with big sizes `type="high"` or both `type="both"`. The `conf` argument specifies the lower and upper confident interval for outliers, the `detect_outlier` is insensistive to the interval values.

```{r, warning=FALSE, message=FALSE}
sce = calculate_QC_metrics(sce)
sce = detect_outlier(sce)
```

we can plot the alignment statistics for the sample data

```{r}
plot_mapping(sce, percentage = TRUE, dataname = "sc_sample_data")
```

and also the pairwise QC metrics plots:

```{r}
plot_QC_pairs(sce)
```

The last step will be to remove the low quality cells by `remove_outliers`. Since the scater and scran package uses the SingleCellExperiment, it will very easy to continue the pipeline with scater and scran, for normalization and visualization.

```{r}
sce = remove_outliers(sce)
dim(sce)
```



