---
title: "scPipe: flexible data preprocessing pipeline for 3' end scRNA-seq data"
author: "Luyi Tian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scPipe: flexible data preprocessing pipeline for 3' end scRNA-seq data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



# Get started

`scPipe` is a package designed to process scRNA-seq data generated by different protocols. It is designed for protocols with UMI, but can also adapt to non-UMI protocols. scPipe consist of two major components. The first is data preprocessing with raw fastq as input and gene counting matrix as output. The second component starts from gene counting matrix, includes quality control and a shiny app for clustering and visualization.

# case study: simulate and process CEL-seq2 data

## simulate a simple CEL-seq2 dataset.

we start from creating a folder to store all results.

```{r}
library(scPipe)
data_dir = "celseq2_demo"
dir.create("celseq2_demo", showWarnings = FALSE)
```

CEL-seq2 is a modified version of CEL-Seq method with higher sensitivity and lower costs. It uses a polyT primer with index sequences to capture cell's mRNA. The index sequence consist of the a cellular specific index, which is designed and the same for each cell, and a molecular index, also called unique molecular identifier (UMI), which is randomly generated and different for each primer. The ouput fastq of Cel-seq2 is paired-ended where the first read is transcript and second read is cellindex+UMI with some polyA sequence.

to simulate a CEL-seq2 dataset. We need the genome fasta file, gff3 exon annotation and a cell barcode annotation. The barcode annotation should be a `.csv` file with at least have two column where first column is the cell id and second column is the barcode sequence. We use ERCC in this demo. All file can be found in `extdata` in this package:

```{r}
# file path:
ERCCfa_fn = system.file("extdata", "ERCC92.fa", package = "scPipe")
ERCCanno_fn = system.file("extdata", "ERCC92_anno.gff3", package = "scPipe")
barcode_annotation_fn = system.file("extdata", "barcode_anno.csv", package = "scPipe")
```

We start from simulating the paired-end fastq files. Read1 contains transcript and Read2 contains UMI followed by cell barcode. For simplixity we use default parameters, more parameter see `?sc_celseq2_simulator`.
```{r}
sc_celseq2_simulator(file.path(data_dir, "simu_R1.fastq"),
                     file.path(data_dir, "simu_R2.fastq"),
                     ERCCanno_fn,
                     barcode_annotation_fn,
                     ERCCfa_fn)
```

### data proprocessing

The pipeline begin with fastq file reformatting. We move the barcode and UMI to the read name and keep the transcript, so the output read name looks like `@[barcode_sequence]*[UMI_sequence]#[readname]` ...

```{r}
sc_trim_barcode(file.path(data_dir, "combined.fastq"),
                file.path(data_dir, "simu_R1.fastq"),
                file.path(data_dir, "simu_R2.fastq"))
```

Then we align the reads to genome. The example uses `Rsubread` but it can be any aligner that support spoice junction alignment.
```{r}
library(Rsubread)
dir.create("celseq2_demo/index")
buildindex(basename="celseq2_demo/index/ERCC_index", reference=ERCCfa_fn)

align(index="celseq2_demo/index/ERCC_index",
      readfile1=file.path(data_dir, "combined.fastq"),
      output_file=file.path(data_dir, "out.aln.bam"), phredOffset=64)
```

After alignment we will map reads to exons using the exon annotation. Currently `scPipe` only supports annotation in `gff3` and `bed` format.

```{r}
sc_exon_mapping(file.path(data_dir, "out.aln.bam"),
                file.path(data_dir, "out.map.bam"),
                ERCCanno_fn)
```

Next we use `sc_demultiplex()` to split the reads to each cell and put them into spearate `.csv` file in `/count` subfolder. Each file contains three column. Each row is a read and first column is the gene is that this read mapped to, second column is the UMI sequence of this read, the third column is the distance to transcript end position. This file will be used for UMI deduplication and to generate gene counting matrix by callling `sc_gene_counting()`.

```{r}
sc_demultiplex(file.path(data_dir, "out.map.bam"), data_dir, barcode_annotation_fn)

sc_gene_counting(data_dir, barcode_annotation_fn)
```

Now we complete the preprocessing step. you will find the gene counting matrix at `data_dir/gene_count.csv` and running statistics in `data_dir/stat` folder, which can be used for quality control.

clean up when we done:
```{r}
unlink(data_dir, recursive = TRUE)
```

## case study: simulate and process Drop-seq data

TODO

## case study: simulate and process MAR-seq data

TODO

## quality control

TODO

## normalization and feature selection

TODO
> give two examples of using BASiC and scran to normalize the data. select high variable genes and coexpressed genes

## dimensionality reduction, clustering and visualization

TODO
> use tSNE and diffusion map. visualization using shiny app. convert to `scater` data format for further analysis
