---
title: "scPipe: flexible data preprocessing pipeline for 3' end scRNA-seq data"
author: "Luyi Tian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scPipe: flexible data preprocessing pipeline for 3' end scRNA-seq data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



# Get started

`scPipe` is a package designed to process scRNA-seq data generated by different protocols. It is designed for protocols with UMI, but can also adapt to non-UMI protocols. scPipe consist of two major components. The first is data preprocessing with raw fastq as input and gene counting matrix as output. The second component starts from gene counting matrix, includes quality control and a shiny app for clustering and visualization.

# case study: simulate and process CEL-seq2 data

## simulate a simple CEL-seq2 dataset.

we start from creating a folder to store all results.

```{r,eval=TRUE}
library(scPipe)
data_dir = "celseq2_demo"
dir.create("celseq2_demo", showWarnings = FALSE)
```

CEL-seq2 is a modified version of CEL-Seq method with higher sensitivity and lower costs. It uses a polyT primer with index sequences to capture cell's mRNA. The index sequence consist of the a cellular specific index, which is designed and the same for each cell, and a molecular index, also called unique molecular identifier (UMI), which is randomly generated and different for each primer. The ouput fastq of Cel-seq2 is paired-ended where the first read is transcript and second read is cellindex+UMI with some polyA sequence.

to simulate a CEL-seq2 dataset. We need the genome fasta file, gff3 exon annotation and a cell barcode annotation. The barcode annotation should be a `.csv` file with at least have two column where first column is the cell id and second column is the barcode sequence. We use ERCC in this demo. All file can be found in `extdata` in this package:

```{r}
# file path:
ERCCfa_fn = system.file("extdata", "ERCC92.fa", package = "scPipe")
ERCCanno_fn = system.file("extdata", "ERCC92_anno.gff3", package = "scPipe")
barcode_annotation_fn = system.file("extdata", "barcode_anno.csv", package = "scPipe")
```

We start from simulating the paired-end fastq files. Read1 contains transcript and Read2 contains UMI followed by cell barcode. For simplixity we use default parameters, more parameter see `?sc_celseq2_simulator`. The read structure for CEL-seq2 is a paired-ended reads with first longer read mapped to transcript and second shorter read consist of 6bp UMI followed by 8bp cell barcode.
```{r,eval=TRUE}
sc_celseq2_simulator(file.path(data_dir, "simu_R1.fastq"),
                     file.path(data_dir, "simu_R2.fastq"),
                     ERCCanno_fn,
                     barcode_annotation_fn,
                     ERCCfa_fn)
```

## data proprocessing

The pipeline begin with fastq file reformatting. We move the barcode and UMI to the read name and keep the transcript, so the output read name looks like `@[barcode_sequence]*[UMI_sequence]#[readname]` ...

```{r,eval=TRUE}
sc_trim_barcode(file.path(data_dir, "combined.fastq"),
                file.path(data_dir, "simu_R1.fastq"),
                file.path(data_dir, "simu_R2.fastq"))
```

Then we align the reads to genome. The example uses `Rsubread` but it can be any aligner that support RNA-seq alignment. The `Rsubread` package does not support Windows platform so the following step will not run on Windows.


```{r,eval=TRUE}
if(!(.Platform$OS.type == "windows")){
  library(Rsubread)
  dir.create("celseq2_demo/index")
  buildindex(basename="celseq2_demo/index/ERCC_index", reference=ERCCfa_fn)

  align(index="celseq2_demo/index/ERCC_index",
      readfile1=file.path(data_dir, "combined.fastq"),
      output_file=file.path(data_dir, "out.aln.bam"), phredOffset=64)
}

```

After alignment we will map reads to exons using the exon annotations. Currently `scPipe` only supports annotation in `gff3` and `bed` format.

```{r,eval=TRUE}
if(!(.Platform$OS.type == "windows")){
  sc_exon_mapping(file.path(data_dir, "out.aln.bam"),
                file.path(data_dir, "out.map.bam"),
                ERCCanno_fn)
}

```

Next we use `sc_demultiplex()` to split the reads to each cell and put them into spearate `.csv` file in `/count` subfolder. Each file contains three column. Each row is a read and first column is the gene is that this read mapped to, second column is the UMI sequence of this read, the third column is the distance to transcript end position. This file will be used for UMI deduplication and to generate gene counting matrix by callling `sc_gene_counting()`.

```{r,eval=TRUE}
if(!(.Platform$OS.type == "windows")){
  sc_demultiplex(file.path(data_dir, "out.map.bam"), data_dir, barcode_annotation_fn,has_UMI=FALSE)

  sc_gene_counting(data_dir, barcode_annotation_fn)
}

```


## do everything in one line of code. 

The tutorial above separate each step, however we also provided a function called `run_scPipe` to integrate them into one step. the `run_scPipe` includes an argument `report` which can be set to `TRUE` to generate an Rmarkdown report with both html report and runnable rmd source code.

```{r,eval=FALSE}
scd = run_scPipe(sample_name="celseq2_demo",
          outdir="celseq2_demo",
          r1=file.path("celseq2_demo", "simu_R1.fastq"),
          r2=file.path("celseq2_demo", "simu_R1.fastq"),
          read_structure=list(bs1 = -1, bl1 = 2, bs2 = 6, bl2 = 8, us = 0, ul = 6),
          filter_settings = list(rmlow = TRUE, rmN = TRUE, minq = 20, numbq = 2),
          genome_index="celseq2_demo/index/ERCC_index",
          exon_anno=ERCCanno_fn,
          barcode_anno=barcode_annotation_fn,
          max_mis=1,
          UMI_cor=1,
          gene_fl=FALSE,
          report=FALSE)
```

Now we complete the preprocessing step. you will find the gene counting matrix at `data_dir/gene_count.csv` and quality control statistics in `data_dir/stat` folder, which can be used for quality control.


# process data generated by other protocols

the most difference that come from different protocols is the read structure, which is defined by `read_structure` argument in scPipe. The difference between CEL-seq and Drop-seq is the cell barcode is unknown for Drop-seq, so an extra step `sc_detect_bc` is required before `sc_demultiplex`, to identify and generate the cell barcode annotation. To use `run_scPipe` on Drop-seq data, just ignore the `barcode_anno` argument. scPipe is not optimized for the full-lengh protocols like SMART-seq, but it can process the data generated by such protocol by set the `has_UMI` to `FALSE` in `sc_demultiplex` or `run_scPipe`.

## quality control

Using the example dataset, which is sampled from Christine Biben's blood cells with different cell type, we use the `detect_outlier` function to find the poor quality cells. The `detect_outlier` has argument `comp` to define the maximum component of the gaussian mixture model. Usually the default value 1 is sufficient, but in case the data is heterogeneous in terms of quality control metrics, the value can be set to 2 or 3 to get a better results. It can only remove low quality cells by `type="low"` or only cells with big sizes `type="high"` or both `type="both"`. The `conf` argument specifies the lower and upper confident interval for outliers, the `detect_outlier` is insensistive to this interval.

```{r}
data("sc_sample_data")
data("sc_sample_qc")
QualityControlInfo = new("AnnotatedDataFrame", data = as.data.frame(sc_sample_qc))
scd = newSCData(countData = as.matrix(sc_sample_data),
               QualityControlInfo = QualityControlInfo,
               useForExprs = "counts",
               organism = "mmusculus_gene_ensembl",
               gene_id_type = "external_gene_name")
scd = calQCMetrics(scd)
scd = detect_outlier(scd)
```

we can plot the alignment statistics for the sample data

```{r}
plot_mapping(scd, percentage = TRUE, dataname = "sc_sample_data")
```

and also the pairwise QC metrics plots:

```{r}
plotQC_pair(scd)
```




