# Detect outliers based on robust linear regression of QQ plot
# @param x a vector of mahalanobis distance
# @param df degree of freedom for chi-square distribution
# @param conf confidence for linear regression
#
# @return cell names of outliers
.qq_outliers_robust <- function(x, df, conf) {
  n <- length(x)
  P <- stats::ppoints(n)
  z <- stats::qchisq(P, df=df)
  ord.x <- x[order(x)]
  coef <- stats::coef(MASS::rlm(ord.x ~ z, maxit=200))
  a <- coef[1]
  b <- coef[2]
  zz <- stats::qnorm(1 - (1 - conf)/2)
  SE <- (b/stats::dchisq(z, df))*sqrt(P*(1 - P)/n)
  fit.value <- a + b*z
  upper <- fit.value + zz*SE
  thr <- max(ord.x[ord.x < upper])
  return(names(x[x > thr]))
}


#' Detect outliers based on QC metrics
#'
#' @description the algorithm will try to find \code{comp} number of components in quality
#' control metrics. Then apply outlier detection on the component with the most genes detected
#' The rest of the components will be considered as poor quality cells. You will get more outlier
#' if you increase the number of \code{comp}.
#' @param sce a \code{SingleCellExperiment} object containing QC metrics in \{colData}.
#' @param comp the number of component used in GMM. can be one or two.
#' @param sel_col a vector of column names, indicate the columns to use for QC.
#' by default it will be the statistics generated by `calQCMetrics()`
#' @param type only looking at low quality cells (`low`) or possible doublets (`high`)
#' or both (`both`)
#' @param conf confidence interval for linear regression at lower and upper tails.
#' Usually this is smaller for lower tail because we hope to pick out more
#' low quality cells than doublets.
#' @param batch whether to perform quality control separately for each batch 
#' default is FALSE. if set to TRUE then you should have a column called `batch` 
#' in the `colData(sce)`
#' 
#' @details detect outlier using Mahalanobis distances
#'
#' @return an updated \code{SingleCellExperiment} object with an `outlier`` column in 
#' \code{int_colData}
#'
#' @importFrom stats cov pchisq mahalanobis complete.cases qchisq
#' @importFrom robustbase covMcd
#' @importFrom mclust Mclust mclustBIC
#'
#' @export
#' @examples
#' data("sc_sample_data")
#' data("sc_sample_qc")
#' QualityControlInfo = new("AnnotatedDataFrame", data = as.data.frame(sc_sample_qc))
#' sce = new\code{SingleCellExperiment}(countData = as.matrix(sc_sample_data),
#'                QualityControlInfo = QualityControlInfo,
#'                useForExprs = "counts",
#'                organism = "mmusculus_gene_ensembl",
#'                gene_id_type = "external_gene_name")
#' sce = calQCMetrics(scd)
#' sce = detect_outlier(scd)
#'
detect_outlier <- function(sce,
                          comp=1,
                          sel_col=NULL,
                          type=c("both", "low", "high"),
                          conf=c(0.9, 0.99),
                          batch=FALSE) {
  type <- match.arg(type)

  # check format:
  if (is(sce, "SingleCellExperiment")) {
    if (is.null(sel_col)) {
      sel_col <- c("number_of_genes", "total_count_per_cell", "non_mt_percent",
                  "non_ERCC_percent")
    }
    if (!all(sel_col %in% colnames(sce@int_colData))) {
      tmp <- sel_col[!(sel_col %in% colnames(sce@int_colData))]
      print("the following QC metrics not find in int_colData from sce:")
      print(tmp)
      if(any(c("number_of_genes", "total_count_per_cell") %in% tmp)){
        stop("the quality control metrics should at least contain the numebr of 
             genes(`number_of_genes`) and counts per cell(`total_count_per_cell`)!")
      }
    }
    x_all <- sce@int_colData[, colnames(sce@int_colData) %in% sel_col]
    x_all$total_count_per_cell = log2(x_all$total_count_per_cell+1)
  }
  else {
    stop("sce must be an `SingleCellExperiment` object.")
  }
  if (!all(complete.cases(x_all))) {
    stop("we find NAs in the selected columns, check the quality control matrix")
  }
  if (is.null(dim(x_all))) {
    sce@int_colData$outliers <- FALSE
    return(sce)
  }
  if(!batch){
    batch_info = rep(1,nrow(x_all))
  }else{
    if ("batch" %in% colnames(colData(sce))){
      batch_info = colData(sce)$batch
    }else{
      stop("did not find column `batch` in colData(sce).")
    }
  }
  outliers = rep(FALSE,nrow(x_all))
  
  for(a_batch in unique(batch_info)){
    x = x_all[batch_info == a_batch, ]
    dist <- mahalanobis(x, center=colMeans(x), cov=cov(x))
    keep <- !(dist>qchisq(0.99, ncol(x)))
    mod <- Mclust(x[keep, ],
                  G=comp,
                  modelNames="EEE",
                  verbose = FALSE)
    if (comp == 1) {
      covr <- covMcd(x, alpha=0.7)
      dist <- mahalanobis(x,
                          center=covr$center,
                          cov=covr$cov)
      mean_diff <- sign(t(x)-covr$center)
      QC_sign <- c(-1, 1)[as.factor(apply(mean_diff, 2, function(t) {sum(t)>0}))]
      neg_dist <- dist[QC_sign == -1]
      pos_dist <- dist[QC_sign == 1]
      if (type == "both") {
        outlier_cells <- .qq_outliers_robust(neg_dist, ncol(x), conf[1])
        outlier_cells <- c(outlier_cells,
                           .qq_outliers_robust(pos_dist, ncol(x), conf[2]))
      }
      else if (type == "low") {
        outlier_cells <- .qq_outliers_robust(neg_dist, ncol(x), conf[1])
      }
      else if (type == "high") {
        outlier_cells <- .qq_outliers_robust(pos_dist, ncol(x), conf[2])
      }
    }
    else {
      ord_fst <- c(1:comp)[order(mod$parameters$mean[1, ], decreasing = TRUE)]
      poor_comp <- ord_fst[2:comp]
      good_comp <- ord_fst[1]
      keep1 <- rep(TRUE, nrow(x))
      keep1[keep][mod$classification %in% poor_comp] <- FALSE
      keep1[!keep] <- FALSE
      sub_x <- x[keep1, ]
      covr <- covMcd(sub_x, alpha=0.7)
      sub_dist <- mahalanobis(sub_x,
                              center=covr$center,
                              cov=covr$cov)
      
      mean_diff <- sign(t(sub_x)-covr$center)
      QC_sign <- c(-1, 1)[as.factor(apply(mean_diff, 2, function(t) {sum(t)>0}))]
      neg_dist <- sub_dist[QC_sign == -1]
      pos_dist <- sub_dist[QC_sign == 1]
      outlier_cells <- .qq_outliers_robust(neg_dist, ncol(sub_x), conf[1])
      outlier_cells <- c(outlier_cells,
                         .qq_outliers_robust(pos_dist, ncol(sub_x), conf[2]))
      outlier_cells <- c(outlier_cells, rownames(x[!keep1, ]))
      if (!(type == "both")) {
        mean_diff <- sign(t(x)-mod$parameters$mean[, good_comp])
        QC_sign <- c(-1, 1)[as.factor(apply(mean_diff, 2, function(t) {sum(t)>0}))]
        if (type == "low") {
          outlier_cells <- rownames(x)[(rownames(x) %in% outlier_cells) & (QC_sign == -1)]
        }
        else if (type == "high") {
          outlier_cells <- rownames(x)[(rownames(x) %in% outlier_cells) & (QC_sign == 1)]
        }
      }
    }
    if(any(rownames(x)  %in% outlier_cells)){
      outliers[batch_info == a_batch][rownames(x)  %in% outlier_cells] = TRUE
    }
  }
  sce@int_colData$outliers = as.factor(outliers)
  return(sce)
}


#' get QC metrics using gene counting matrix
#'
#' @param sce an \code{SingleCellExperiment} object containing count
#' @details get QC metrics using gene counting matrix
#' the QC statistics added are
#' `number_of_genes`: number of genes detected
#' `total_count_per_cell`: sum of read number after UMI dedup
#' `non_mt_percent`: 1- percent of mitochondrial gene counts
#' mitochondrial genes are retrived by GO term GO:0005739
#' `non_ERCC_percent`: ratio of exon counts to ERCC counts
#' `non_ribo_percent`: 1- percent of ribosomal gene counts
#' ribosomal genes are retrived by GO term GO:0005840
#' @return an \code{SingleCellExperiment} with updated QC metrics
#'
#' @importFrom Biobase exprs pData fData
#'
#' @export
#' @examples
#'
calQCMetrics <- function(sce) {
  if (!is(scd, "SingleCellExperiment")) {
    stop("require a `SingleCellExperiment` object.")
  }else if(!("counts" %in% names(assays(sce)))){
    stop("counts not in names(assays(sce)). cannot find count data.")
  }
  # get gene number
  gene_number <- colSums(assay(sce,"counts")>0)
  if (all(gene_number == 0)) {
    stop("all genes have zero count. check your expression matrix.")
  }else{
    sce@int_colData$number_of_genes = gene_number
  }
  
  # get total count per cell
  sce@int_colData$total_count_per_cell = colSums(assay(sce,"counts"))
  
  # get ERCC ratio
  if(!is.null(spikeNames(sce))){
    if(any(isSpike(sce,"ERCC"))){
      exon_count = colSums(assay(sce,"counts")[!isSpike(sce,"ERCC"),])
      ERCC_count = assay(sce,"counts")[isSpike(sce,"ERCC"),]
      sce@int_colData$non_ERCC_percent = exon_count/(ERCC_count+exon_count+1e-5)
    }else{
      print("no ERCC spike-in. skip `non_ERCC_percent`")
    }
  }else{
    print("spikeNames(sce)==NULL, no spike-in information.")
  }

  # get mt percentage
  if (!(gene_id_type(scd) == "NA")) {
    mt_genes <- get_genes_by_GO(returns=gene_id_type(scd),
                                dataset=organism.\code{SingleCellExperiment}(scd),
                                go=c("GO:0005739"))
    if (length(mt_genes)>0) {
      if (sum(rownames(exprs_mat) %in% mt_genes)>1) {
        mt_count <- colSums(exprs_mat[rownames(exprs_mat) %in% mt_genes, ])
        QC_met$non_mt_percent <- (exon_count-mt_count)/(exon_count+0.01) # add 0.01 to make sure they are not NA
      }
    }
  }
  else {
    print("no gene_id_type, skip `non_mt_percent`")
  }

  # get ribosomal percentage
  if (!(gene_id_type(scd) == "NA")) {
    ribo_genes <- get_genes_by_GO(returns=gene_id_type(scd),
                               dataset=organism.\code{SingleCellExperiment}(scd),
                               go=c("GO:0005840"))
    if (length(ribo_genes)>0) {
      if (sum(rownames(exprs_mat) %in% ribo_genes)>1) {
        ribo_count <- colSums(exprs_mat[rownames(exprs_mat) %in% ribo_genes, ])
        QC_met$non_ribo_percent <- (exon_count-ribo_count)/(exon_count+0.01)
      }
    }
  }
  else {
    print("no gene_id_type, skip `non_ribo_percent`")
  }
  QCMetrics(scd) <- QC_met
  return(scd)
}


#' plot QC statistics for \code{SingleCellExperiment} object
#' @param sce an \code{SingleCellExperiment} object
#' @param sel_col a vector of column names, indicate the columns to use for plot.
#' by default it will be the statistics generated by `calQCMetrics()`
#' @importFrom GGally ggpairs
#' @import ggplot2
#' @export
#' @return a ggplot2 object
#' @examples 
#' data("sc_sample_data")
#' data("sc_sample_qc")
#' QualityControlInfo = new("AnnotatedDataFrame", data = as.data.frame(sc_sample_qc))
#' sce = new\code{SingleCellExperiment}(countData = as.matrix(sc_sample_data),
#'                QualityControlInfo = QualityControlInfo,
#'                useForExprs = "counts",
#'                organism = "mmusculus_gene_ensembl",
#'                gene_id_type = "external_gene_name")
#' sce = calQCMetrics(scd)
#' sce = detect_outlier(scd)
#' plotQC_pair(scd)
#' 
plotQC_pair <- function(scd, sel_col=NULL) {
  if (is(scd, "\code{SingleCellExperiment}")) {
    if (is.null(sel_col)) {
      sel_col <- c("number_of_genes", "total_count_per_cell", "non_mt_percent",
                   "non_ERCC_percent", "non_ribo_percent", "outliers")
    }
    x <- pData(QCMetrics(scd))[, colnames(QCMetrics(scd)) %in% sel_col]
  }
  else {
    stop("sce must be an \code{SingleCellExperiment} object.")
  }
  if ("outliers" %in% colnames(x)) {
    return(ggpairs(x, mapping = ggplot2::aes_string(colour = "outliers")))
  }
  else {
    return(ggpairs(x))
  }
}

#' plot mapping statistics for \code{SingleCellExperiment} object.
#' @param sce an \code{SingleCellExperiment} object
#' @param sel_col a vector of column names, indicate the columns to use for plot.
#' by default it will be the mapping result.
#' @param percentage logic. whether to convert the number of reads to percentage
#' @param dataname the name of this dataset which appears in the plot title
#'
#' @import scales reshape ggplot2
#' @importFrom stats prcomp reorder
#' @return a ggplot2 object
#' @export
#' 
#' @examples 
#' data("sc_sample_data")
#' data("sc_sample_qc")
#' QualityControlInfo = new("AnnotatedDataFrame", data = as.data.frame(sc_sample_qc))
#' sce = new\code{SingleCellExperiment}(countData = as.matrix(sc_sample_data),
#'                QualityControlInfo = QualityControlInfo,
#'                useForExprs = "counts",
#'                organism = "mmusculus_gene_ensembl",
#'                gene_id_type = "external_gene_name")
#' sce = calQCMetrics(scd)
#' sce = detect_outlier(scd)
#' plot_mapping(scd,percentage=TRUE,dataname="sc_sample")
#'
plot_mapping <- function(scd,
                        sel_col=NULL,
                        percentage=FALSE,
                        dataname="") {
  if (is(scd, "\code{SingleCellExperiment}")) {
    if (is.null(sel_col)) {
      sel_col <- c("unaligned", "aligned_unmapped", "ambiguous_mapping",
                   "mapped_to_ERCC", "mapped_to_intron", "mapped_to_exon")
    }
    x <- pData(QCMetrics(scd))[, sel_col]
  }
  else {
    stop("sce must be an \code{SingleCellExperiment} object.")
  }


  mapping_stat <- x
  mapping_stat$sample_name <- stats::reorder(rownames(mapping_stat), mapping_stat$mapped_to_exon)
  mapping_stat_prop <- as.data.frame(prop.table(as.matrix(mapping_stat[, sapply(mapping_stat, is.numeric)]), 1))
  mapping_stat_prop$sample_name <- mapping_stat$sample_name
  dat.m <- melt(mapping_stat, id.vars="sample_name")
  dat.m1 <- melt(mapping_stat_prop, id.vars="sample_name")

  if (!percentage) {
    p <- ggplot(dat.m, aes_string(x="sample_name", y="value", fill="variable")) + scale_fill_brewer(palette="Set1")+
      geom_bar(stat="identity", width=1)+
      ylab("number of reads")+
      xlab("cell sorted by number of reads mapped to exon")+
      theme(axis.title.x=element_blank(), axis.text.x=element_blank())+
      ggtitle(paste0("overall mapping statistics of ", dataname, " (number of reads)"))
  }
  else {
    p <- ggplot(dat.m1, aes_string(x="sample_name", y="value", fill="variable")) + scale_fill_brewer(palette="Set1")+
      geom_bar(stat="identity", width=1)+
      ylab("percentage of reads")+
      xlab("cell sorted by number of reads mapped_to_exon")+
      scale_y_continuous(labels=percent_format())+
      theme(axis.title.x=element_blank(), axis.text.x=element_blank())+
      ggtitle(paste0("overall mapping statistics of ", dataname, " (percentage)"))
  }

  return(p)
}


#' remove outliers for \code{SingleCellExperiment}
#' @param sce an \code{SingleCellExperiment} object
#' @export
#' @return an \code{SingleCellExperiment} object without outliers
#' @examples 
#' data("sc_sample_data")
#' data("sc_sample_qc")
#' QualityControlInfo = new("AnnotatedDataFrame", data = as.data.frame(sc_sample_qc))
#' sce = new\code{SingleCellExperiment}(countData = as.matrix(sc_sample_data),
#'                QualityControlInfo = QualityControlInfo,
#'                useForExprs = "counts",
#'                organism = "mmusculus_gene_ensembl",
#'                gene_id_type = "external_gene_name")
#' sce = calQCMetrics(scd)
#' sce = detect_outlier(scd)
#' scd_afterqc = remove_outliers(scd)
#' 
remove_outliers <- function(scd) {
  if (!is(scd, "SingleCellExperiment")) {
    stop("sce must be an \code{SingleCellExperiment} object.")
  }
  if (!("outliers" %in% colnames(pData(QCMetrics(scd))))) {
    stop("no outlier information. please run `detect_outlier()` first.")
  }
  out_cell <- pData(QCMetrics(scd))$outliers == FALSE
  return(scd[, out_cell])
}
