


#' create SCData object from data folder generated by preprocessing step
#' 
#' after we run \code{sc_gene_counting} and finish the preprocessing step. \code{create_scd_by_dir}
#' can be used to generate the SCData obeject from the folder that contains gene counting matrix and QC statistics.
#'
#' @param datadir the directory that contains all the data and `stat` subfolder.
#' @param species the organism of the data. List of possible names can be retrieved using the function 
#' `listDatasets`from `biomaRt` package. (i.e `mmusculus_gene_ensembl` or `hsapiens_gene_ensembl`)
#' @param gene_id gene id type of the data A possible list of ids can be retrieved using the function `listAttributes` from `biomaRt` package. 
#' the commonly used id types are `external_gene_name`, `ensembl_gene_id` or `entrezgene`
#' @param pheno_data the external phenotype data that linked to each single cell. this should be an \code{AnnotatedDataFrame} object
#' 
#' @details after we run \code{sc_gene_counting} and finish the preprocessing step. \code{create_scd_by_dir}
#' can be used to generate the SCData obeject from the folder that contains gene counting matrix and QC statistics.
#' 
#' @return an SCData object
#'
#'
#' @export
#' 

create_scd_by_dir = function(datadir, species=NULL, gene_id=NULL, pheno_data=NULL){
  gene_cnt = read.csv(file.path(datadir, "gene_count.csv"),row.names=1)
  cell_stat = read.csv(file.path(datadir,"stat","cell_stat.csv"),row.names=1)
  
  gene_cnt = gene_cnt[,order(colnames(gene_cnt))]
  cell_stat = cell_stat[order(rownames(cell_stat)),]
  
  
  QualityControlInfo = new("AnnotatedDataFrame", data = as.data.frame(cell_stat))
  scd = newSCData(countData = as.matrix(gene_cnt),
                  QualityControlInfo = QualityControlInfo,
                  phenoData = pheno_data,
                  useForExprs = "counts",
                  organism = species,
                  gene_id_type = gene_id)
  
  return(scd)
}


#' create_report
#' 
#' create html report using data generated by proprocessing step.
#'
#' @param sample_name sample name
#' @param outdir output folder
#' @param r1 file path of read1
#' @param r2 file path of read2 default to be NULL
#' @param outfq file path of the output of \code{sc_trim_barcode}
#' @param read_structure a list contains read structure configuration. for more help see `?sc_trim_barcode`
#' @param filter_settings a list contains read filter settings for more help see `?sc_trim_barcode`
#' @param align_bam the aligned bam file
#' @param genome_index genome index used for alignment
#' @param map_bam the mapped bam file
#' @param exon_anno the gff exon annotation used
#' @param stnd whether to perform strand specific mapping
#' @param fix_chr add `chr` to chromosome names, fix inconsistant names.
#' @param barcode_anno genome annotation, gff file. can have multiple files
#' @param max_mis maximum mismatch allowed in barcode. default to be 1
#' @param UMI_cor correct UMI sequence error: 0 means no correction, 1 means simple correction and merge UMI with distance 1.
#' @param gene_fl whether to remove low abundant gene count. low abundant is defined as only one copy of one UMI for this gene
#' @param species the organism of the data. List of possible names can be retrieved using the function 
#' `listDatasets`from `biomaRt` package. (i.e `mmusculus_gene_ensembl` or `hsapiens_gene_ensembl`)
#' @param gene_id_type gene id type of the data A possible list of ids can be retrieved using the function `listAttributes` from `biomaRt` package. 
#' the commonly used id types are `external_gene_name`, `ensembl_gene_id` or `entrezgene`
#'
#' @return
#' @export
#'
#' @examples
create_report = function(sample_name,
                         outdir,
                         r1,
                         r2=NULL,
                         outfq,
                         read_structure,
                         filter_settings,
                         align_bam,
                         genome_index,
                         map_bam,
                         exon_anno,
                         stnd,
                         fix_chr,
                         barcode_anno,
                         max_mis,
                         UMI_cor,
                         gene_fl,
                         species,
                         gene_id_type){
  SAMPLE_NAME=sample_name
  
  FQ1=r1
  FQ2=r2
  FQOUT=outfq
  if(read_structure$bs1<0){
    BC1_INFO="NA"
  }else{
    BC1_INFO = paste0("start at position ",read_structure$bs1,", length ",read_structure$bl1)
  }
  
  
  BC2_INFO=paste0("start at position ",read_structure$bs2,", length ",read_structure$bl2)
  UMI_INFO=paste0("start at position ",read_structure$us,", length ",read_structure$ul)
  
  RM_N=as.character(filter_settings$rmN)
  RM_LOW=as.character(filter_settings$rmlow)
  MIN_Q=filter_settings$minq
  NUM_BQ=filter_settings$numbq
  
  BAM_ALIGN=align_bam
  G_INDEX=genome_index
  
  BAM_MAP=map_bam
  
  OUTDIR=outdir
  
  ANNO_GFF=exon_anno
  
  STND=as.character(stnd)
  FIX_CHR=as.character(fix_chr)
  BC_ANNO=barcode_anno
  MAX_MIS=max_mis
  
  if(UMI_cor == 1){
    UMI_COR="simple correction and merge UMI with distance 1"
  }else if(UMI_cor == 0){
    UMI_COR="no correction"
  }else{
    UMI_COR="unknown"
  }
  
  
  
  GENE_FL=as.character(gene_fl)
  
  SPECIES=species
  GENE_ID_TYPE=gene_id_type
  
  rmarkdown::render(system.file("extdata", "report_template.Rmd", package = "scPipe"), output_file = file.path(outdir,"report.html"))
}